tests/generative/README.md

```diff
@@ -0,0 +1,425 @@
+@hyp.given(variable=variable_strategy, data=data_strategy)
+@hyp.settings(**settings)
+def test_query_model(query_engines, variable, data, recorder):
+    recorder.record_inputs(variable, data)  # this is used to record and report on some helpful data about the tests
+    run_test(query_engines, data, variable, recorder)
+```
+
+## data strategy
+
+We define a simplified table schema that contains columns of various types. The [`data_setup`](data_setup.py) uses this schema to setup up a number of patient and event tables (2 of each), and
+the [`data_strategy`](data_strategies.py) populates the tables with hypothesis-generated data
+for each test.
+
+## value strategies
+
+The value strategies are strategies defining simple types; `int`, `bool`, `date`, `float` and `str`.
+
+```
+value_strategies = {
+    int: st.integers(min_value=0, max_value=10),
+    bool: st.booleans(),
+    datetime.date: st.dates(
+        min_value=datetime.date(2010, 1, 1), max_value=datetime.date(2020, 12, 31)
+    ),
+    float: st.floats(min_value=0.0, max_value=11.0, width=16, allow_infinity=False),
+    str: st.text(alphabet=["a", "b", "c"], min_size=0, max_size=3),
+}
+```
+
+The same value strategies are used for defining both data strategies and variable (query model)
+strategies. Generally, we try to define these values with narrow enough ranges that they will
+sometimes overlap, and we can test equality. e.g. if we are testing an addition operation that
+takes 2 ints, we can be reasonably sure that `st.integers(min_value=0, max_value=10)` will
+test addition of two ints that are the same at some point.
+
+## variable strategies
+
+The variable strategies are the most complex part of databuilder's generative test strategies.
+
+A variable is defined by calling `variable()` in [`variable_strategies.py`](variable_strategies.py),
+with the tables, schema and value strategies as described above.
+
+`variable` defines lots of inner functions, each of which returns a strategy for creating the
+thing it is named for, not the thing itself.
+For example, the `value` inner function returns a strategy for creating `Value` objects, and not a `Value` object itself.
+
+A valid variable is a `Series`, chosen by selecting a type (one of the types in `value_strategies`) and a frame
+that the variable must be consistent with (with one row per patient, because we require that any
+variable on a dataset represents one row per patient). The variable returned will be a
+series of the chosen type, consistent with the chosen frame.
+
+```
+def variable(patient_tables, event_tables, schema, value_strategies):
+   ...
+
+    @st.composite
+    def valid_variable(draw):
+        type_ = draw(any_type())
+        frame = draw(one_row_per_patient_frame())
+        return draw(series(type_, frame))
+
+    return valid_variable()
+```
+
+Let's take a closer look at the elements of the `valid_variable`.
+
+`type_` is a strategy for choosing a type, sampled from the types used as the keys for
```

Comment:
This isn't quite right. `any_type()` returns the strategy you describe. `type_` is an actual type, drawn from the strategy. (The same applies to `frame` below.)